lucene/core/src/java/org/apache/lucene/util/crypto/Crypto.java

```diff
@@ -42,31 +47,71 @@
   public static int AES_KEY_SIZE = 256;
   public static int AES_BLOCK_SIZE = 16;
   public static String CTR_TRANSFORM = "AES/CTR/NoPadding";
+  private static Path INDICES_PATH = Paths.get("indices");
 
   public static volatile AtomicBoolean encryptionOn = new AtomicBoolean(true);
-
-  public static SecretKey TEST_AES_KEY = new SecretKeySpec(
-      Base64.getDecoder().decode("4tZ9S+gRYX2F3fm+BIWDDvkcXbkKYXBmB27hixPvSjU="), AES_ALGORITHM);
-
+  public static volatile AtomicBoolean testingOn = new AtomicBoolean(false);
+
+  public static List<SecretKey> TEST_AES_KEYS =  Arrays.asList(
+      new SecretKeySpec(
+          Base64.getDecoder().decode("4tZ9S+gRYX2F3fm+BIWDDvkcXbkKYXBmB27hixPvSjU="), AES_ALGORITHM),
+      new SecretKeySpec(
+          Base64.getDecoder().decode("7C9Q6SnqXwrfI03LPE2J0LEcnnVP/YnF8O3hGvpjs7Q="), AES_ALGORITHM),
+      new SecretKeySpec(
+          Base64.getDecoder().decode("UqVyK+sHZQVGJBbTr8UDN5TF/t5f9MgqmVkSxi6uJu0="), AES_ALGORITHM),
+      new SecretKeySpec(
+          Base64.getDecoder().decode("pd0fWwS7glXPpbzrPavNx7zpo25DLXa1gAbsijz2pHw="), AES_ALGORITHM),
+      new SecretKeySpec(
+          Base64.getDecoder().decode("dtCrOKsvBlgAc3ZAGAxlWHlHK3dNNXjhrCtjXXJavtk="), AES_ALGORITHM)
+  );
+  
   public static boolean setEncryptionOn(boolean on) {
     return encryptionOn.getAndSet(on);
   }
 
   public static boolean isEncryptionOn() {
     return encryptionOn.get();
   }
+  
+  public static boolean setTestingOn(boolean on) {
+    return testingOn.getAndSet(on);
+  }
+
+  public static boolean isTestingOn() {
+    return testingOn.get();
+  }
 
   public static void initialize() throws NoSuchAlgorithmException {
     Security.setProperty("crypto.policy", "unlimited");
     getSecureRandom();
   }
+  
+  // made public for testing purposes
+  public static String getIndexUid(Path path) {
+    Iterator<Path> pathIterator = path.iterator();
+    while (pathIterator.hasNext()) {
+      if (pathIterator.next().equals(INDICES_PATH)) {
+        // next segment will contain the index UUID
+        return pathIterator.next().toString();
+      }
+    }
+    if (isTestingOn()) {
```

Comment:
You don't usually add a branch in the main code specifying what to do during the tests. You should mock the method and its behavior in the tests themselves.

Reply:
I'm not sure how to do this yet - probably some `KeyProvider` singleton should be created and passed to `Crypto`, normally it will access the StrongDoc API but during testing it will be replaced by a mocked version.
This will also decouple `Crypto` from future key-fetching-from-StrongDoc logic.
(Let me know if you have better ideas)

Reply:
This approach is ok if you know that we will need multiple key provider implementations (not including the mock one). But if it's designed like this just so you can use some mock implementation it looks like an unnecessarily complicated design to me.

What I initially had in mind was using some testing/mocking library for that. They are already using [Mockito](https://site.mockito.org/) in their tests. You can simply say I want **this** method call to return **this** value during the testing ([https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#2](https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#2)).

But feel free to keep it this way if you think it's better. I'm not entirely convinced either way myself :smile:.

Reply:
Ah, ok. I expected that some reflection / aspects would need to be implemented (from scratch) to be able to do that. I'll keep it as-is for now and maybe refactor it with mockito later...